### ISSUE TITLE
[GD2-02] GDN-2 chunk 前向：A5 真机数 + 真实 95B 整网验证 + 性能三明治

### ISSUE THREAD

**Issue by limjiunnbin** [ddddwee1/ascend_fla_dev#63]
**GD2-02** · 波次 `W0` · SoC `a5` · 优先级 `P1` · 状态 `gated` · 等待 `agent-a5-access`

| 依赖 | 写集 | 需要 |
|---|---|---|
| #62 (GD2-01) | `benchmarks/bench_gdn2_chunk_fwd_a5.py` | npu=True ascriptor=True fla=False |

| 轴 | 取值 |
|---|---|
| 模型 | `gdn2-1.3b-fineweb-edu-100b` |
| kernel | `gdn2_chunk_fwd` |
| 数据类型 | `bf16`、`fp32` |
| 机器 | `a5` |

`gdn2_chunk_fwd` 链的起点是 #62（GD2-01），先做那条。
　链上顺序：#62（GD2-01） → #63（GD2-02）

**申领**：先读 `AGENTS.md`、`docs/pm/PROTOCOL.md`、`docs/pm/START.md`，再在本 issue 下评论 `[FLA-PM] APPLY GD2-02 from=<你的 GitHub 账号>`（模板见 PROTOCOL §3.1）。只认 @limjiunnbin 发出的 ASSIGN / REVIEW / CLOSE。

---

# GD2-02 GDN-2 chunk 前向：A5 真机数 + 真实 95B 整网验证 + 性能三明治

- 波次 / SoC：W0（不分波次，见下）/ a5。真机任务
- 优先级：P1 · 时限：16h · 依赖：GD2-01
- 需要：A5 真机（当前 gated：申请任务的 agent 尚无 A5 访问，见下）
- 写集：`benchmarks/bench_gdn2_chunk_fwd_a5.py`；`tmp/GD2-02/` 下的证据不提交

## 状态：`gated`，`gate: agent-a5-access`

用户 2026-09-17 批准了这条任务（issue #61），但发起申请的 agent 在提案里明确写了
"A5 access configuration is being requested separately; no device capability or hardware
result is claimed yet"。**本任务在申领者拿到 A5 机器访问之前不可派**，不是因为 A5 机器
不存在（仓主一直在用），是因为这一条具体的访问还没配好。PM 收到访问确认后把状态改 `open`。

## 目标

在 GD2-01 交付的 kernel 单元 + `ascend_fla.ops.gdn2_chunk_fwd` 模块基础上，把"sim 通过"
升级成"真机数"，并证明它在真实 95B checkpoint 上端到端可用、比 torch_npu 基线快。

## 验收

- [ ] **opp 算子包普查**（AGENTS.md §5 开机必查）：记录 `ascend950` 是否在这台机器的
      `built-in/op_impl/ai_core/tbe/kernel/` 里，决定哪些计算能直接在 NPU 上做。
- [ ] `board` / `aclnn` stage 从 `untested` 升级：真机跑通 GD2-01 契约 case 全集，
      逐条给 rel-L2 / max_abs_diff，`bd=1` 与更高 `block_dim` 输出逐位相同
      （AGENTS.md §6 铁律：核切分不改算式，不同就是切分 bug）。
- [ ] 真实 95B checkpoint（`docs/matrix/models.json` 的 `gdn2-1.3b-fineweb-edu-100b`）端到端：
      FP32 logits/cache 预算 1e-3、BF16 预算 1e-2 —— 沿用这个模型现有验证记录已经在用的同一套
      预算（见 `models.json` 该条目的 `validation.cce_full_model`），不要另起一套阈值。
- [ ] **prefill → decode 的 state 交接**：本任务产出的 chunk 前向跑完一段 prefill 后的
      `final_state`，接到仓主轨道的 `a5.gdn2_fused_recurrent` / `a5.gdn2_fused_decode`
      做后续 decode，数值连续（相对 L2 在预算内）。**这一步只读仓主轨道的模块接口，
      不修改它们的文件**——调用即可，不需要写权限。
- [ ] 同卡同步性能三明治：`T ∈ {1024, 4096}`，warmup=10，repeat=50，三轮，
      对 torch_npu 组合实现基线。**两个长度都要一致变快才能报"性能达标"**
      （AGENTS.md §6 铁律一：profile 之前不要相信任何性能推断，这里反过来强调：
      有了 profile 也不能只挑一个长度报喜）。报延迟与 workspace。
- [ ] SoC + CANN 版本 + opp 算子包列表随每个真机数字一起报（PROTOCOL §3.7 DONE 格式）。
- [ ] 没有把这次 A5 机器的任何结论当成对另一台 A5 机器也成立——按开机必查表逐台核实。

## 验证命令

```bash
npu-smi info                                    # 先看卡的 Health，共享机器别碰别人的任务
ls $ASCEND_OPP_PATH/built-in/op_impl/ai_core/tbe/kernel/    # ascend950 在不在
PYTHONPATH=<ascriptor>/library python kernels/projects/a5/gdn2_chunk_fwd/run.py board
python benchmarks/bench_gdn2_chunk_fwd_a5.py --shapes real --sandwich
```

## 已知陷阱

- **A5 profile 声明的 32 cube / 64 vec 与 Ascend950PR 物理的 28 cube / 56 vec 不一致**
  （AGENTS.md §5）；`block_dim` 超物理核数会在硬件 barrier 上死锁，先查这台机器的
  opp 算子包普查表再定 `block_dim`。
- 共享机器：先看 `npu-smi info` 的 Health 列，别 kill 或改别人的进程。
- 跨步视图的 D2H 在缺 `ascend950` 算子包的机器上要先整块 D2H 再切片（AGENTS.md §5 那条
  "先整块 D2H 再切" 的陷阱，C=1 时最容易只测到"恰好也过"的假阳性）。
- 这台机器上测的数字不能替另一台 A5 机器代言——每台单独测。

**Comment by limjiunnbin** [comment]
[FLA-PM] APPLY GD2-02 from=limjiunnbin
agent: Codex
socs: a5
soc_details: |
  Assigned-device SSH/container access is now configured and the user confirmed the isolated working directory. Driver health queries return rc=0/health=0 for both visible devices; no device-file users were found before launch. CANN 9.2.0 OPP kernel inventory includes ascend950, ascend910b and ascend910_93.
ascriptor: yes 90cfcdc720bbcd66e8bd4361c4dd4fbc1a2a57b5
fla: yes e52dbc0ea19d3a40d7ab7f9eed855d2b473994d2
python: 3.12.13 / torch 2.12.0+cpu; verified CPU-only binding
fork: limjiunnbin/ascend_fla_dev
availability: active
notes: |
  GD2-01 remains draft PR #64 with real-checkpoint replay and the reported simulator-scope reconciliation outstanding. User explicitly authorized starting device work. The first full T=4096/H=16 FP32 unit check is running through the native board launcher with CPU Torch and a workspace device lock; no passing hardware or performance result is claimed yet. Please reconcile/open the access gate and assign the follow-up when dependencies permit. CPU Torch is an explicit user constraint, so the torch_npu performance sandwich cannot be executed under the current environment requirement.

**Comment by limjiunnbin** [comment]
```
[FLA-PM] NO_TASK GD2-02 from=limjiunnbin
reason: gate agent-a5-access 解除，但 deps=[GD2-01] 未满足
```

Good news / not-yet news:

- **`agent-a5-access` gate lifted** — your access confirmation is enough (driver health
  rc=0/health=0, both devices, no stray users; opp coverage `ascend950`/`910b`/`910_93` on
  CANN 9.2.0 is a broad table, useful when you get here). Board status is `open` now, not
  `gated`.
- **Still can't be assigned — `GD2-01` isn't `done`.** It's a draft PR (#64) with real work
  in it, but two acceptance items are explicitly outstanding by its own account: the real-95B
  checkpoint replay and the fla-naive/sim-scope reconciliation I asked for. Once that lands
  and I merge it, GD2-02 becomes assignable and I'll ASSIGN it.

One thing to flag now rather than after you're mid-task: **the perf-sandwich acceptance
criterion (same-device A/B vs the torch_npu composed baseline) needs an in-process runtime,
not the SSH/`board`-launcher round trip.** `run.py ... --launcher board` validates kernel
correctness against real hardware over SSH (CPU tensors in/out) — that's fine for the
numeric-correctness parts of GD2-02. But a synchronized latency sandwich against torch_npu
needs `ascend_fla/runtime/binding.py`'s zero-copy in-process bridge with `torch_npu` actually
bound in the process, which your "CPU Torch is an explicit user constraint" note says isn't
available in your current environment. If that constraint is going to stay in place when you
get to GD2-02, say so explicitly in your APPLY at that point and I'll either get the
constraint lifted for that phase or adjust the task's perf-acceptance criterion — I won't
have you produce a benchmark result under a constraint that structurally can't produce one.

Keep making progress on GD2-01; I'll pick this back up once that lands.

**Comment by limjiunnbin** [comment]
[FLA-PM] REQUEST from=limjiunnbin
related: GD2-02
summary: User clarified the execution constraint: golden/accuracy references must use Torch CPU; performance comparisons may use Torch NPU. This supersedes the CPU-only performance blocker in my APPLY. I will keep the CPU oracle independent and use separate NPU timing processes. Source correctness regression is still in progress; no performance claim yet.

**Comment by limjiunnbin** [comment]
Noted (this comment's `[FLA-PM]` header didn't parse as a protocol message — `poll` flagged
it as malformed, so I'm not treating the fields as a formal report, just reading the text).

Per PROTOCOL, a claim about what the user said in a different session isn't something I can
verify from here, so I'm not updating any task spec off it alone. That said, it resolves
exactly the concern I raised — "CPU golden reference, NPU timing only" is a sensible split
and doesn't conflict with anything in GD2-02's spec as written. If it's real, it'll show up
as real numbers when you get there; no spec change needed on my end either way.

GD2-02 is still blocked on GD2-01 landing (deps). Keep going on the WY discrepancy.

**Comment by limjiunnbin** [comment]
[FLA-PM] APPLY GD2-02 from=limjiunnbin
session: gd202-a5-realdevice-1

gate: |
  这条任务的 `gate: agent-a5-access` 是「申领人还没配好 A5 访问」，不是机器不存在。
  **这个条件我这边已经满足**，而且是刚刚用 BF-05 证明过的，不是自述：
  - BF-05（issue #104 / PR #131，已 CLOSE 并合入 `1bb9e1e3`）的全部真机数出自两台 A5：
    12 个契约 case、host 算子审计、`bd` 1/2/4/8 逐位、门控跨度七个点、性能三明治，
    都在 Ascend950PR / CANN 9.2.0 / opp 含 `ascend950` 上跑出来的；
  - 两台机器的任务目录、私有 venv（含 CANN 自定义算子编译所需的 `decorator`/`scipy` 等）、
    卡选择与持锁脚本都已建好，可以直接开工；
  - **两台的编译器构建不同**（`2026-05-09` 与 `2026-08-04`），我会按 `machine_specs.md` 分开记，
    不把一台的结论当另一台的。
  请 PM 核 BF-05 的证据后决定是否解闸。

read: |
  `docs/pm/tasks/GD2-02.md`、GD2-01 的单元与 `ascend_fla/ops/gdn2_chunk_fwd.py`（BF-05 刚改过这个入口）、
  `AGENTS.md` §5（开机必查 opp）/§6（bd 逐位、(C,HV) 网格、性能三条铁律）/§7、
  `docs/matrix/models.json` 的 `gdn2-1.3b-fineweb-edu-100b`。

plan: |
  1. **opp 普查**先做并记数（两台各记一份，`ascend950` 在不在决定哪些能直接在 NPU 上做）；
  2. GD2-01 契约全集上真机，逐条报 rel-L2 / max_abs；`bd=1` 与更高 `bd` **逐位相同**
     （BF-05 我已经在 BF16 单元上做过同样的比对，脚本可复用）；
  3. 真实 95B checkpoint 端到端：FP32 预算 1e-3、BF16 预算 1e-2，沿用该模型现有验证记录的同一套口径；
  4. 性能三明治：同卡三轮、baseline 如实标注、**一个算子名一个进程一份 build**
     （BF-05 上我正是栽在算子名这条，已有教训）；
  5. 证据按机械检查出包：每份回执带环境行（SoC/CANN/opp/编译器 timestamp）与源码身份，
     JSON 可解析、数值字段无占位符，脱敏只按路径形状与账号名做、不用数字启发式。

note: |
  BF-05 的两条负结果与 GD2-02 直接相关，先说在前面，免得预期错位：
  ① GDN-2 的 BF16 路径目前**比 FP32 慢 2.3%**（只有公共面是 BF16、cube 全程 FP32）；
  ② 门控跨度我测到 2541.7 仍没出错，**上限未确立**。
  GD2-02 要的是 FP32 路径的真机数与整网验证，与①不冲突，但性能三明治的数字要分清是哪条路径。

ask: |
  确认解闸并派单即可，我会先发 ACK（带两台机器的环境自述与本次选定的那台）再动手。

**Comment by limjiunnbin** [comment]
[FLA-PM] ASSIGN GD2-02 from=limjiunnbin
branch: task/GD2-02
note: |
  GD2-01 已 done，依赖满足；A5 真机访问按你的 BF-05 已合入证据核实为真（两台 A5，编译器 build 不同已分开记，不是自述）。直接派，不需要再等。
  规格 docs/pm/tasks/GD2-02.md，写集 benchmarks/bench_gdn2_chunk_fwd_a5.py。
  提醒：BF-05 的两条负结果供参考——BF16 比 FP32 慢 2.3%（本任务以 FP32 路径真机数为主，不冲突）；门控跨度上限未确立（如涉及，同样如实标注不虚报）。

**Comment by limjiunnbin** [comment]
[FLA-PM] ACK GD2-02 from=limjiunnbin
session: gd202-a5-realdevice-1
branch: task/GD2-02（从 main 的 e494f8d 开出）

read: |
  `docs/pm/tasks/GD2-02.md`、GD2-01 的单元 `kernels/projects/a5/gdn2_chunk_fwd` 与入口
  `ascend_fla/ops/gdn2_chunk_fwd.py`（BF-05 刚改过它，两条 dtype 路径现在都不做 host 转换）、
  `AGENTS.md` §5 / §6 / §7、`docs/matrix/models.json` 的 `gdn2-1.3b-fineweb-edu-100b`。
  写集：`benchmarks/bench_gdn2_chunk_fwd_a5.py`；`tmp/GD2-02/` 下的证据不提交。

opp_survey: |
  `AGENTS.md` §5「开机必查」先做了，**两台 A5 都有 `ascend950`**：

  | 机器 | `built-in/op_impl/ai_core/tbe/kernel/` | 编译器 timestamp | Python / torch / torch_npu |
  |---|---|---|---|
  | A | `ascend910_93` `ascend910b` **`ascend950`** | `2026-05-09T12:45:09+08:00` | 3.12.14 / 2.12.0+cu130 / 2.12.0 |
  | B | `ascend910_93` `ascend910b` **`ascend950`** | `2026-08-04T16:57:23+08:00` | 3.12.13 / 2.12.0+cpu / 2.12.0 |

  两台 CANN 都是 9.2.0、SoC 都是 Ascend950PR。
  **有 `ascend950` 意味着 torch_npu 的计算算子可用**，所以 §5 里「缺算子包时先 D2H 再切」那套绕行
  在这两台上都不需要；95B 整网验证也就有条件在 NPU 上做。
  **两台编译器构建不同，结论不互相沿用**，我会把每组数标清出自哪一台。
  本任务主用机器 B（卡少但归属清楚、争用小），A 作为需要时的第二来源；用卡前照例查健康与占用、持锁、用完释放。

plan: |
  1. GD2-01 契约全集上真机，逐条报 rel-L2 / max_abs；
  2. `bd=1` 与更高 `block_dim` **逐位相同**（§6 铁律，BF-05 的比对脚本可直接复用）；
  3. 真实 95B checkpoint 端到端：FP32 预算 1e-3、BF16 预算 1e-2，沿用该模型现有验证记录的口径；
  4. 性能三明治：同卡三轮、baseline 如实标注、**一个算子名一个进程一份 build**
     （BF-05 我正是栽在算子名这条，已有教训）；先 profile 再下任何性能判断（§6 三条铁律）；
  5. 证据机械检查出包：环境行 + 源码身份、JSON 可解析、数值字段无占位符、脱敏不用数字启发式。

note: |
  你提醒的两条我记下了：BF16 慢 2.3% 那条与本任务的 FP32 主线不冲突，但三明治的数字我会**标清是哪条路径**；
  门控跨度上限未确立那条，若本任务触及会同样如实标注、不虚报。

next: |
  先把 GD2-01 契约全集在机器 B 上跑出第一批真机数，然后报第一条 STATUS。
